src/software/jetson_nano/services/network/network.cpp

```diff
@@ -74,3 +77,66 @@ void NetworkService::primitiveSetCallback(TbotsProto::PrimitiveSet input)
                      << "%";
     }
 }
+
+void NetworkService::logNewPrimitiveSet(TbotsProto::PrimitiveSet input)
+{
+    /* TODO: THIS TRACKS THE RECEIVED PRIMITIVES
+     *  - make sure it is in the lock
+     *  - log the new PRIMITIVE SET PROTO within a deque if it is valid (see primitive_tracker.send())
+     *  - update the timestamp on each primitive set to the current epoch time
+     */
+
+    if (primitive_set_tuple_rtt.size() >= 50)
+    {
+        LOG(WARNING) << "Too many primitive sets logged for round-trip calculations, halting log process";
+        return;
+    }
+
+    if (!primitive_set_tuple_rtt.empty()
+            && input.sequence_number() <= std::get<0>(primitive_set_tuple_rtt.back()))
+    {
+        // If the proto is older than the last received proto, then ignore it
+        return;
+    }
+    primitive_set_tuple_rtt.emplace_back(
+            std::tuple(input.sequence_number(),
+                       getCurrentEpochTimeInSeconds(),
+                       input.time_sent().epoch_timestamp_seconds())
+                       );
+}
+
+void NetworkService::updatePrimitiveSetLog(TbotsProto::RobotStatus &robot_status)
+{
+    /**
+     * TODO: Send return RobotStatus msg to Thunderscope
+     *      1) Traverse deque containing all received primitive_sets
+     *      2) If the last_handled_primitive_set for current robot_status is stored in the deque:
+     *         - delete all earlier sets
+     *         - stop & calculate the processing counter
+     *         - store the new omit_thunderloop_processing_time_sent = time_sent + processing duration
+     */
+
+    uint64_t seq_num = robot_status.last_handled_primitive_set();
+    while (seq_num <= std::get<0>(primitive_set_tuple_rtt.back()))
+    {
+        if (std::get<0>(primitive_set_tuple_rtt.front()) == seq_num)
+        {
+            double received_epoch_time_seconds = std::get<1>(primitive_set_tuple_rtt.front());
+            double processing_time_seconds = getCurrentEpochTimeInSeconds() - received_epoch_time_seconds;
+
+            robot_status.mutable_omit_thunderloop_processing_time_sent()->set_epoch_timestamp_seconds(
+                    std::get<2>(primitive_set_tuple_rtt.front())
+                    + processing_time_seconds
+            );
+            return;
+        }
+        primitive_set_tuple_rtt.pop_front();
+    }
+    LOG(WARNING) << "No primitives available for round-trip calculations";
+}
+
+double NetworkService::getCurrentEpochTimeInSeconds()
+{
+    return std::chrono::duration<double>
+            (std::chrono::steady_clock::now().time_since_epoch()).count();
+}
```

Comment:
Yes, it does get the time in seconds. I have no idea why `std::chrono` is so convoluted in its type conversions, but this time is a duration for how long the jetson has been powered on. This is because I could not figure out how to have the jetson connect to the internet to sync its onboard time (I am also not sure if this is possible as we would always be connected to the tbots network). In application, since I am using the delta of the times, it makes no difference. TLDR: Yes I will reformat it to the suggested form